navigator/runtime-experimental/src/test/kotlin/com/freeletics/mad/navigator/compose/internal/MultiStackTest.kt

```diff
@@ -0,0 +1,736 @@
+        stack.push(SimpleRoute(3))
+        stack.push(SimpleRoute(4))
+        stack.push(SimpleRoute(5))
+        stack.push(OtherRoute(6))
+        stack.push(ThirdRoute(7))
+        stack.push(OtherRoute(8))
+        stack.push(OtherRoute(9))
+        stack.push(ThirdRoute(10))
+
+        assertThat(stack.visibleEntries.value)
+            .containsExactly(
+                StackEntry(StackEntry.Id("104"), SimpleRoute(5), simpleRouteDestination),
+                StackEntry(StackEntry.Id("105"), OtherRoute(6), otherRouteDestination),
+                StackEntry(StackEntry.Id("106"), ThirdRoute(7), thirdRouteDestination),
+                StackEntry(StackEntry.Id("107"), OtherRoute(8), otherRouteDestination),
+                StackEntry(StackEntry.Id("108"), OtherRoute(9), otherRouteDestination),
+                StackEntry(StackEntry.Id("109"), ThirdRoute(10), thirdRouteDestination),
+            )
+            .inOrder()
+        assertThat(stack.canNavigateBack.value).isTrue()
+
+        assertThat(removed).isEmpty()
+    }
+
+    @Test
+    fun `visibleEntries is all entries starting from last screen 2`() {
+        val stack = underTest()
+        stack.push(SimpleRoute(2))
+        stack.push(SimpleRoute(3))
+        stack.push(SimpleRoute(4))
+        stack.push(OtherRoute(5))
+        stack.push(ThirdRoute(6))
+        stack.push(SimpleRoute(7))
+        stack.push(OtherRoute(8))
+        stack.push(OtherRoute(9))
+        stack.push(ThirdRoute(10))
+
+        assertThat(stack.visibleEntries.value)
+            .containsExactly(
+                StackEntry(StackEntry.Id("106"), SimpleRoute(7), simpleRouteDestination),
+                StackEntry(StackEntry.Id("107"), OtherRoute(8), otherRouteDestination),
+                StackEntry(StackEntry.Id("108"), OtherRoute(9), otherRouteDestination),
+                StackEntry(StackEntry.Id("109"), ThirdRoute(10), thirdRouteDestination),
+            )
+            .inOrder()
+        assertThat(stack.canNavigateBack.value).isTrue()
+
+        assertThat(removed).isEmpty()
+    }
+
+    @Test
+    fun `visibleEntries is new entry after navigating to root`() {
+        val stack = underTest()
+        stack.push(OtherRoot(1), restoreRootState = false, saveCurrentRootState = false)
+
+        assertThat(stack.visibleEntries.value)
+            .containsExactly(
+                StackEntry(StackEntry.Id("101"), OtherRoot(1), otherRootDestination)
+            )
+            .inOrder()
+        assertThat(stack.canNavigateBack.value).isTrue()
+
+        assertThat(removed).isEmpty()
+    }
+
+    @Test
+    fun `visibleEntries is new entry after navigating to root twice`() {
+        val stack = underTest()
+        stack.push(OtherRoot(1), restoreRootState = true, saveCurrentRootState = true)
+        stack.push(SimpleRoot(1), restoreRootState = false, saveCurrentRootState = false)
+        stack.push(OtherRoot(1), restoreRootState = true, saveCurrentRootState = true)
```

Comment:
I think this API is hard to follow and understand end result. Maybe having separate methods for each use case would make it clearer. This is kind of related to general question if we need to support permutations of boolean parameters. What do you think?

Reply:
Thanks for the explanation.

> So we never have restoreRootState = true, saveCurrentRootState: Boolean = false. We could make resetToRoot a separate API which is the case where both are false

I like that, I think it would be good to have it this way.

Reply:
Split it up here [30ad593](https://github.com/freeletics/mad/pull/372/commits/30ad5932e5316346f1387f5adf70b5fc8db304b8)